compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/codegeneration/BackendCodeGenerator.kt

```diff
@@ -321,38 +337,23 @@ fun ProgramNode.compileToKotlin(
             codeGenerators,
             protocolComposer
         )
-
-        val classBuilder = TypeSpec.classBuilder(
-            host.name.replaceFirstChar { it.uppercase() }
-        ).primaryConstructor(
-            FunSpec.constructorBuilder()
-                .addParameter("runtime", Runtime::class)
-                .build()
-        ).addProperty(
-            PropertySpec.builder("runtime", Runtime::class)
-                .initializer("runtime")
-                .build()
-        )
-
-        for (function in this.functions) {
-            classBuilder.addFunction(
-                curGenerator.generateFunction(
-                    host,
-                    function.name.value.name,
-                    function
-                )
-            ).build()
-        }
-        objectBuilder.addType(classBuilder.build())
+        val hostTypeSpec = curGenerator.generateClass()
+        objectBuilder.addType(curGenerator.generateClass())
+        hostClassMap[host] = hostTypeSpec
     }
 
+    // create main function
+    val mainFunctionBuilder = FunSpec.builder("main").addModifiers(KModifier.SUSPEND)
+    mainFunctionBuilder.addParameter("host", Host::class)
+    mainFunctionBuilder.addParameter("runtime", Runtime::class)
+
     // create switch statement in main method so program can be run on any host
     mainFunctionBuilder.beginControlFlow("when (host)")
     for (host in this.hosts) {
         mainFunctionBuilder.addStatement(
             "%N -> %N(%L).main()",
             host.name,
-            host.name.replaceFirstChar { it.uppercase() },
+            hostClassMap[host]!!.name!!,
```

Comment:
Final style thing. Don't extract `name` manually, KotlinPoet will do it for you. Also, when accessing a map, instead of `map[key]!!`, use `map.getValue(key)`. It will give a better error message.